File: problems/112/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        bool ans = false;
        
        dfs(root, targetSum, ans);
        
        return ans;
    }
    
     void dfs(TreeNode* root, int sum, bool& ans) {
         if (not root or ans) {
             return;
         }
         
         sum -= root->val;
         
         if (not root->left and not root->right and not sum) {
             ans = true;
             return;
         }
         
         dfs(root->left, sum, ans);
         dfs(root->right, sum, ans);
     }
};

File: problems/112/solution.go
package main
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func hasPathSum(root *TreeNode, sum int) bool {
    if root == nil {
        return false
    }
    
    sum -= root.Val
    
    if root.Left == nil && root.Right == nil && sum == 0 {
        return true
    }
    
    return hasPathSum(root.Left, sum) || hasPathSum(root.Right, sum)
}